statics/js/topologymanager_package/src/topologymanager/data_source/websocket.ts

```diff
@@ -0,0 +1,82 @@
+import * as events from 'events';
+
+export default class WSHandler {
+    protocol: string;
+    e: events.EventEmitter = new events.EventEmitter();
+    conn: any;
+    connectionUrl: string;
+    connecting: boolean = false;
+    connected: boolean = false;
+    reconnect: boolean = true;
+
+    constructor(connectionUrl: string) {
+        this.connectionUrl = connectionUrl;
+        this.protocol = "ws://";
+        if (location.protocol == "https:") {
+            this.protocol = "wss://";
+        }
+    }
+
+    connect() {
+        if (this.conn && this.conn.readyState == window.WebSocket.OPEN) {
+            return;
+        }
+        if (this.connecting || this.connected) {
+            return;
+        }
+        this._connect();
+    }
+
+    _connect() {
+        this.connecting = true;
+        this.connected = false;
+        this.conn = new WebSocket(this.protocol + this.connectionUrl);
+        this.conn.onopen = () => {
+            this.connecting = false;
+            this.connected = true;
+            this.e.emit('websocket.connected');
+        };
+        this.conn.onclose = () => {
+            this.e.emit('websocket.disconnected');
+            this.connected = false;
+            this.tryToReconnect();
+        };
+        this.conn.onmessage = (r: any) => {
+            const msg = JSON.parse(r.data);
+            this.e.emit('websocket.message' + msg.Namespace, msg);
+        };
+        this.conn.onerror = () => {
+            this.e.emit('websocket.error');
+            this.connected = false;
+            this.tryToReconnect();
+        };
+    }
+
+    disconnect() {
+        this.connected = false;
+        this.conn && this.conn.close();
+    }
+
+    send(msg: any) {
+        this.conn.send(JSON.stringify(msg));
+    }
+
+    tryToReconnect() {
+        if (!this.reconnect) {
+            return;
+        }
+        this.conn.onopen = function() { }
```

Comment:
I think there is an issue here. Once reconnected you will miss the events and then won't resync the graph

Reply:
> I think there is an issue here. Once reconnected you will miss the events and then won't resync the graph

I don't like this part as well. But it works. About reconnecting -> I tested and it works. But yes you are right it is a suspicious part. Unfortunately it's impossible to make the code perfect :) So if it works, then it's ok :) And most important it is wrapped in one class. So, later we can easily fix it.